### PR TITLE
Fix/csi 1552 fu 1339 create snap ds8k

### PR DESCRIPTION
**Pull request by ArbelNathan**
follow up for -
https://github.com/IBM/ibm-block-csi-driver/pull/163

### CODE REVIEW

File: controller/array_action/array_mediator_ds8k.py [hunk 1]
```diff
@@ -52,8 +52,15 @@ def parse_version(bundle):
     return '.'.join([v1[-1], v2[0], v2[1]])
 
 
-def get_volume_id_from_scsi_identifier(scsi_id):
-    return scsi_id[-4:]
+def get_volume_id_from_scsi_identifier(func):
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:55]: nice!
here is how I would write it:
```
from decorator import decorator

VOLUME_ID_PARAM_NAME = 'volume_id'

def scsi_id_to_volume_id(scsi_id):
    return scsi_id[-4:]

def try_convert_first_arg(converter, args):
    if args:
        converted = converter(args[0])
        return (converted,) + args[1:]
    return ()

def try_convert_arg_by_name(converter, name, kwargs):
    if name in kwargs:
        kwargs[name] = converter(kwargs[name])

@decorator
def convert_volume_id_from_scsi_id(func, *args, **kwargs):
    args = try_convert_first_arg(scsi_id_to_volume_id, args)
    try_convert_arg_by_name(scsi_id_to_volume_id, VOLUME_ID_PARAM_NAME, kwargs)
    return func(*args, **kwargs)

@convert_volume_id_from_scsi_id
def is_flashcopy_source(volume_id, volume_flashcopy):
    ....
```
feel free to take whatever you like from it

File: controller/array_action/array_mediator_ds8k.py [hunk 2]
```diff
@@ -217,20 +224,20 @@ def create_volume(self, name, size_in_bytes, capabilities, pool_id):
...
-                volume = self.client.create_volume(**cli_kwargs)
+            # get the volume before creating again, to make sure it is not existing,
+            # because volume name is not unique in ds8k.
+            api_volume = self._get_api_volume_by_name(
+                name,
+                pool_id=pool_id
+            )
+            logger.info("Found volume {}".format(name))
+            if api_volume is not None:
+                return api_volume
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:235]: shouldn't it be `return self._generate_volume_response(api_volume)`?
if so, please covert this with a unit test

File: controller/array_action/array_mediator_ds8k.py [hunk 3]
```diff
@@ -64,9 +71,9 @@ def get_source_volume_id_if_exists(api_volume):
     return flashcopy_rel_sources[0]
 
 
+@get_volume_id_from_scsi_identifier
 def is_flashcopy_source(volume_id, volume_flashcopy):
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:75]: remove dead code

File: controller/array_action/array_mediator_ds8k.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_ds8k.py:55]: but its not always "volume_id"

Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py:55]: right. then consider this:
```
def convert_from_scsi_id(scsi_id_param_name):
    @decorator
    def convert_to_object_id(func, self, *args, **kwargs):
        args = try_convert_first_arg(scsi_id_to_volume_id, args)
        if not args:
            try_convert_arg_by_name(scsi_id_to_volume_id, scsi_id_param_name, kwargs)
        return func(self, *args, **kwargs)
    return convert_to_object_id

@convert_from_scsi_id('volume_id')
def delete_volume(self, volume_id):
    ....
```

Comment (ArbelNathan) [reply controller/array_action/array_mediator_ds8k.py]: done

Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py]: just don't forget to cover the kwargs use case in the unit tests as well

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_ds8k.py]: would `mediator_method` be better?

[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_ds8k.py]: mediator_method is what we are doing.. what about target_method?

File: controller/array_action/array_mediator_ds8k.py [hunk 4]
```diff
@@ -294,9 +309,9 @@ def _delete_volume(self, volume_id, not_exist_err=True):
             )
             raise array_errors.VolumeDeletionError(volume_id)
 
+    @convert_from_scsi_id()
```
Comment (oriyarde) [inline controller/array_action/array_mediator_ds8k.py:312]: are you sure `()` is needed?

Comment (ArbelNathan) [reply controller/array_action/array_mediator_ds8k.py:312]: yep

Comment (ArbelNathan) [reply controller/array_action/array_mediator_ds8k.py:312]: something with the mediator module..

**Pull request merged**